File: go/internal/util/evaluate.go
package util

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

// EvaluationMetrics contains computed performance metrics from task execution.
type EvaluationMetrics struct {
	TaskSuccessRate        float64 `json:"task_success_rate"`
	FirstAttemptSuccessRate float64 `json:"first_attempt_success_rate"`
	AvgAttempts            float64 `json:"avg_attempts"`
	TokensPerTask          int     `json:"tokens_per_task"`
	CostPerTask            float64 `json:"cost_per_task"`
	QualityPassRate        float64 `json:"quality_pass_rate"`
	FunctionalPassRate     float64 `json:"functional_pass_rate"`
	TestEdgeCaseRate       float64 `json:"test_edge_case_rate"`
	CompletedCount         int     `json:"completed_count"`
	FailedCount            int     `json:"failed_count"`
	TotalTokens            int     `json:"total_tokens"`
	TotalCostUSD           float64 `json:"total_cost_usd"`
	CriteriaPass           int     `json:"criteria_pass"`
	CriteriaPartial        int     `json:"criteria_partial"`
	CriteriaFail           int     `json:"criteria_fail"`
	TotalCriteria          int     `json:"total_criteria"`
	QualityDimensions      map[string]int `json:"quality_dimensions"`
	QualityTotals          map[string]int `json:"quality_totals"`
}

// FailedTask represents a task that failed during execution.
type FailedTask struct {
	ID    string `json:"id"`
	Name  string `json:"name"`
	Error string `json:"error"`
}

// ImprovementPattern describes a pattern identified for improvement.
type ImprovementPattern struct {
	Description string `json:"description"`
	Count       int    `json:"count"`
}

// EvaluationResult contains the full evaluation report data.
type EvaluationResult struct {
	PlanningVerdict      string               `json:"planning_verdict"`
	PlanningIssuesCount  int                  `json:"planning_issues_count"`
	Metrics              *EvaluationMetrics   `json:"metrics"`
	FailedTasks          []FailedTask         `json:"failed_tasks"`
	ImprovementPatterns  []ImprovementPattern `json:"improvement_patterns"`
	TotalTasks           int                  `json:"total_tasks"`
	BlockedCount         int                  `json:"blocked_count"`
	SkippedCount         int                  `json:"skipped_count"`
}

// stateTask matches the task structure in state.json for evaluation purposes.
type stateTask struct {
	ID           string                 `json:"id"`
	Name         string                 `json:"name"`
	Status       string                 `json:"status"`
	Attempts     int                    `json:"attempts"`
	Error        string                 `json:"error"`
	Verification map[string]interface{} `json:"verification"`
}

// stateExecution matches the execution structure in state.json.
type stateExecution struct {
	CompletedCount int     `json:"completed_count"`
	FailedCount    int     `json:"failed_count"`
	TotalTokens    int     `json:"total_tokens"`
	TotalCostUSD   float64 `json:"total_cost_usd"`
}

// stateForEval is a minimal state structure for evaluation.
type stateForEval struct {
	Tasks     map[string]stateTask   `json:"tasks"`
	Execution stateExecution         `json:"execution"`
	Artifacts map[string]interface{} `json:"artifacts"`
}

// RunEvaluation runs a full evaluation on the state file in the planning directory
// and returns the evaluation result with computed metrics.
func RunEvaluation(planningDir string) (*EvaluationResult, error) {
	statePath := filepath.Join(planningDir, "state.json")
	data, err := os.ReadFile(statePath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("state file not found: %s", statePath)
		}
		return nil, fmt.Errorf("failed to read state file: %w", err)
	}

	var state stateForEval
	if err := json.Unmarshal(data, &state); err != nil {
		return nil, fmt.Errorf("failed to parse state JSON: %w", err)
	}

	metrics := computeMetrics(&state)
	failedTasks := getFailedTasks(&state)
	patterns := getImprovementPatterns(&state)

	blockedCount := 0
	skippedCount := 0
	for _, task := range state.Tasks {
		if task.Status == "blocked" {
			blockedCount++
		}
		if task.Status == "skipped" {
			skippedCount++
		}
	}

	planningVerdict := "N/A"
	planningIssuesCount := 0
	if artifacts, ok := state.Artifacts["task_validation"].(map[string]interface{}); ok {
		if verdict, ok := artifacts["verdict"].(string); ok {
			planningVerdict = verdict
		}
		if issues, ok := artifacts["issues"].([]interface{}); ok {
			planningIssuesCount = len(issues)
		}
	}

	return &EvaluationResult{
		PlanningVerdict:     planningVerdict,
		PlanningIssuesCount: planningIssuesCount,
		Metrics:             metrics,
		FailedTasks:         failedTasks,
		ImprovementPatterns: patterns,
		TotalTasks:          len(state.Tasks),
		BlockedCount:        blockedCount,
		SkippedCount:        skippedCount,
	}, nil
}

func computeMetrics(state *stateForEval) *EvaluationMetrics {
	completed := state.Execution.CompletedCount
	failed := state.Execution.FailedCount
	totalTokens := state.Execution.TotalTokens
	totalCost := state.Execution.TotalCostUSD
	totalFinished := completed + failed

	firstAttemptSuccesses := 0
	qualityFullPass := 0
	totalCriteria := 0
	criteriaPass := 0
	criteriaPartial := 0
	criteriaFail := 0
	tasksWithTests := 0
	edgeCasesPass := 0
	totalAttempts := 0

	qualityDimensions := map[string]int{"types": 0, "docs": 0, "patterns": 0, "errors": 0}
	qualityTotals := map[string]int{"types": 0, "docs": 0, "patterns": 0, "errors": 0}

	for _, task := range state.Tasks {
		if task.Status == "complete" {
			attempts := task.Attempts
			if attempts == 0 {
				attempts = 1
			}
			totalAttempts += attempts
			if attempts == 1 {
				firstAttemptSuccesses++
			}

			if task.Verification != nil {
				allQualityPass := true
				if quality, ok := task.Verification["quality"].(map[string]interface{}); ok {
					for _, dim := range []string{"types", "docs", "patterns", "errors"} {
						if score, ok := quality[dim].(string); ok {
							qualityTotals[dim]++
							if score == "PASS" {
								qualityDimensions[dim]++
							} else {
								allQualityPass = false
							}
						}
					}
				}
				if allQualityPass {
					qualityFullPass++
				}

				if criteria, ok := task.Verification["criteria"].([]interface{}); ok {
					for _, c := range criteria {
						if crit, ok := c.(map[string]interface{}); ok {
							totalCriteria++
							if score, ok := crit["score"].(string); ok {
								switch score {
								case "PASS":
									criteriaPass++
								case "PARTIAL":
									criteriaPartial++
								default:
									criteriaFail++
								}
							}
						}
					}
				}

				if tests, ok := task.Verification["tests"].(map[string]interface{}); ok {
					tasksWithTests++
					if edgeCases, ok := tests["edge_cases"].(string); ok && edgeCases == "PASS" {
						edgeCasesPass++
					}
				}
			}
		}
	}

	metrics := &EvaluationMetrics{
		CompletedCount:    completed,
		FailedCount:       failed,
		TotalTokens:       totalTokens,
		TotalCostUSD:      totalCost,
		CriteriaPass:      criteriaPass,
		CriteriaPartial:   criteriaPartial,
		CriteriaFail:      criteriaFail,
		TotalCriteria:     totalCriteria,
		QualityDimensions: qualityDimensions,
		QualityTotals:     qualityTotals,
	}

	if totalFinished > 0 {
		metrics.TaskSuccessRate = float64(completed) / float64(totalFinished)
	}
	if completed > 0 {
		metrics.FirstAttemptSuccessRate = float64(firstAttemptSuccesses) / float64(completed)
		metrics.AvgAttempts = float64(totalAttempts) / float64(completed)
		metrics.TokensPerTask = totalTokens / completed
		metrics.CostPerTask = totalCost / float64(completed)
		metrics.QualityPassRate = float64(qualityFullPass) / float64(completed)
	}
	if totalCriteria > 0 {
		metrics.FunctionalPassRate = float64(criteriaPass) / float64(totalCriteria)
	}
	if tasksWithTests > 0 {
		metrics.TestEdgeCaseRate = float64(edgeCasesPass) / float64(tasksWithTests)
	}

	return metrics
}

func getFailedTasks(state *stateForEval) []FailedTask {
	var failed []FailedTask
	for id, task := range state.Tasks {
		if task.Status == "failed" {
			errorMsg := task.Error
			if errorMsg == "" {
				errorMsg = "Unknown error"
			}
			failed = append(failed, FailedTask{
				ID:    id,
				Name:  task.Name,
				Error: errorMsg,
			})
		}
	}
	return failed
}

func getImprovementPatterns(state *stateForEval) []ImprovementPattern {
	var patterns []ImprovementPattern
	edgeCaseIssues := 0
	typeIssues := 0
	docIssues := 0

	for _, task := range state.Tasks {
		if task.Verification != nil {
			if tests, ok := task.Verification["tests"].(map[string]interface{}); ok {
				if edgeCases, ok := tests["edge_cases"].(string); ok {
					if edgeCases == "PARTIAL" || edgeCases == "FAIL" {
						edgeCaseIssues++
					}
				}
			}

			if quality, ok := task.Verification["quality"].(map[string]interface{}); ok {
				if types, ok := quality["types"].(string); ok {
					if types == "PARTIAL" || types == "FAIL" {
						typeIssues++
					}
				}
				if docs, ok := quality["docs"].(string); ok {
					if docs == "PARTIAL" || docs == "FAIL" {
						docIssues++
					}
				}
			}
		}
	}

	if edgeCaseIssues > 0 {
		patterns = append(patterns, ImprovementPattern{
			Description: fmt.Sprintf("%d task(s) had issues with edge case testing", edgeCaseIssues),
			Count:       edgeCaseIssues,
		})
	}
	if typeIssues > 0 {
		patterns = append(patterns, ImprovementPattern{
			Description: fmt.Sprintf("%d task(s) had type annotation issues", typeIssues),
			Count:       typeIssues,
		})
	}
	if docIssues > 0 {
		patterns = append(patterns, ImprovementPattern{
			Description: fmt.Sprintf("%d task(s) had documentation issues", docIssues),
			Count:       docIssues,
		})
	}

	return patterns
}

// FormatEvaluationReport formats an EvaluationResult as a human-readable string.
func FormatEvaluationReport(result *EvaluationResult) string {
	var output string

	output += "Execution Evaluation Report\n"
	output += "============================================================\n\n"

	output += "Planning Quality\n"
	output += "----------------------------------------\n"
	output += fmt.Sprintf("Plan Verdict: %s\n", result.PlanningVerdict)
	output += fmt.Sprintf("Issues at Planning: %d\n\n", result.PlanningIssuesCount)

	output += "Execution Summary\n"
	output += "----------------------------------------\n"
	output += fmt.Sprintf("Tasks: %d total\n", result.TotalTasks)
	if result.Metrics.CompletedCount > 0 {
		pct := float64(result.Metrics.CompletedCount) / float64(result.TotalTasks) * 100
		output += fmt.Sprintf("  Completed:     %d (%.0f%%)\n", result.Metrics.CompletedCount, pct)
	}
	if result.Metrics.FailedCount > 0 {
		pct := float64(result.Metrics.FailedCount) / float64(result.TotalTasks) * 100
		output += fmt.Sprintf("  Failed:        %d (%.0f%%)\n", result.Metrics.FailedCount, pct)
	}
	if result.BlockedCount > 0 {
		pct := float64(result.BlockedCount) / float64(result.TotalTasks) * 100
		output += fmt.Sprintf("  Blocked:       %d (%.0f%%)\n", result.BlockedCount, pct)
	}
	if result.SkippedCount > 0 {
		pct := float64(result.SkippedCount) / float64(result.TotalTasks) * 100
		output += fmt.Sprintf("  Skipped:       %d (%.0f%%)\n", result.SkippedCount, pct)
	}
	output += "\n"

	firstAttemptCount := int(result.Metrics.FirstAttemptSuccessRate * float64(result.Metrics.CompletedCount))
	output += fmt.Sprintf("First-Attempt Success: %d/%d (%.0f%%)\n",
		firstAttemptCount, result.Metrics.CompletedCount,
		result.Metrics.FirstAttemptSuccessRate*100)
	output += fmt.Sprintf("Average Attempts: %.2f\n\n", result.Metrics.AvgAttempts)

	output += "Verification Breakdown\n"
	output += "----------------------------------------\n"
	output += "Functional Criteria:\n"
	if result.Metrics.TotalCriteria > 0 {
		passRate := float64(result.Metrics.CriteriaPass) / float64(result.Metrics.TotalCriteria) * 100
		partialRate := float64(result.Metrics.CriteriaPartial) / float64(result.Metrics.TotalCriteria) * 100
		failRate := float64(result.Metrics.CriteriaFail) / float64(result.Metrics.TotalCriteria) * 100
		output += fmt.Sprintf("  PASS:     %d/%d (%.0f%%)\n", result.Metrics.CriteriaPass, result.Metrics.TotalCriteria, passRate)
		output += fmt.Sprintf("  PARTIAL:  %d/%d (%.0f%%)\n", result.Metrics.CriteriaPartial, result.Metrics.TotalCriteria, partialRate)
		output += fmt.Sprintf("  FAIL:     %d/%d (%.0f%%)\n", result.Metrics.CriteriaFail, result.Metrics.TotalCriteria, failRate)
	} else {
		output += "  No criteria data\n"
	}
	output += "\n"

	output += "Code Quality:\n"
	for _, dim := range []string{"types", "docs", "patterns", "errors"} {
		total := result.Metrics.QualityTotals[dim]
		passed := result.Metrics.QualityDimensions[dim]
		if total > 0 {
			dimTitle := dim
			if len(dim) > 0 {
				dimTitle = string(dim[0]-32) + dim[1:]
			}
			output += fmt.Sprintf("  %-10s %d/%d PASS\n", dimTitle, passed, total)
		}
	}
	output += "\n"

	output += "Cost Analysis\n"
	output += "----------------------------------------\n"
	output += fmt.Sprintf("Total Tokens:  %d\n", result.Metrics.TotalTokens)
	output += fmt.Sprintf("Total Cost:    $%.2f\n", result.Metrics.TotalCostUSD)
	if result.Metrics.CompletedCount > 0 {
		output += fmt.Sprintf("Per Task:      $%.4f\n", result.Metrics.CostPerTask)
	}
	output += "\n"

	if len(result.FailedTasks) > 0 {
		output += "Failure Analysis\n"
		output += "----------------------------------------\n"
		for _, ft := range result.FailedTasks {
			output += fmt.Sprintf("%s: FAIL - %s\n", ft.ID, ft.Error)
		}
		output += "\n"
	}

	if len(result.ImprovementPatterns) > 0 {
		output += "Improvement Patterns\n"
		output += "----------------------------------------\n"
		for _, p := range result.ImprovementPatterns {
			output += fmt.Sprintf("- %s\n", p.Description)
		}
		output += "\n"
	}

	return output
}

File: scripts/evaluate.py
#!/usr/bin/env python3
"""
Evaluation Report Generator for Task Decomposition Protocol v2

Generates comprehensive performance reports from execution state.
Answers: "How well did tasker perform this job?"

Usage:
    evaluate.py                    Generate full evaluation report
    evaluate.py --format json      Output as JSON
    evaluate.py --metrics-only     Output only computed metrics
    evaluate.py --validate         Include validation checks (DAG, steel thread, etc.)
"""

import json
import sys
from pathlib import Path

# Paths relative to script location
SCRIPT_DIR = Path(__file__).resolve().parent
PROJECT_ROOT = SCRIPT_DIR.parent
TASKER_DIR = PROJECT_ROOT / ".tasker"
STATE_FILE = TASKER_DIR / "state.json"

# Import validation functions
try:
    from validate import (
        compute_calibration_metrics,
        run_all_validations,
        validate_all_verification_commands,
        validate_dag,
        validate_steel_thread,
    )
    VALIDATION_AVAILABLE = True
except ImportError:
    VALIDATION_AVAILABLE = False


def load_state() -> dict | None:
    """Load state from file or return None if doesn't exist."""
    if not STATE_FILE.exists():
        return None
    return json.loads(STATE_FILE.read_text())


def compute_metrics(state: dict) -> dict:
    """Compute performance metrics from state."""
    tasks = state.get("tasks", {})
    execution = state.get("execution", {})

    completed = execution.get("completed_count", 0)
    failed = execution.get("failed_count", 0)
    total_tokens = execution.get("total_tokens", 0)
    total_cost = execution.get("total_cost_usd", 0.0)

    total_finished = completed + failed

    first_attempt_successes = 0
    quality_full_pass = 0
    total_criteria = 0
    criteria_pass = 0
    criteria_partial = 0
    criteria_fail = 0
    tasks_with_tests = 0
    edge_cases_pass = 0
    total_attempts = 0

    # Quality dimension counters
    quality_dimensions = {"types": 0, "docs": 0, "patterns": 0, "errors": 0}
    quality_totals = {"types": 0, "docs": 0, "patterns": 0, "errors": 0}

    for task in tasks.values():
        if task.get("status") == "complete":
            attempts = task.get("attempts", 1)
            total_attempts += attempts
            if attempts == 1:
                first_attempt_successes += 1

            verification = task.get("verification", {})

            # Quality pass rate
            quality = verification.get("quality", {})
            if quality:
                all_pass = True
                for dim in ["types", "docs", "patterns", "errors"]:
                    score = quality.get(dim)
                    if score:
                        quality_totals[dim] += 1
                        if score == "PASS":
                            quality_dimensions[dim] += 1
                        else:
                            all_pass = False
                if all_pass:
                    quality_full_pass += 1

            # Functional criteria pass rate
            criteria = verification.get("criteria", [])
            for c in criteria:
                total_criteria += 1
                score = c.get("score")
                if score == "PASS":
                    criteria_pass += 1
                elif score == "PARTIAL":
                    criteria_partial += 1
                else:
                    criteria_fail += 1

            # Test edge case rate
            tests = verification.get("tests", {})
            if tests:
                tasks_with_tests += 1
                if tests.get("edge_cases") == "PASS":
                    edge_cases_pass += 1

    return {
        "task_success_rate": completed / total_finished if total_finished > 0 else 0.0,
        "first_attempt_success_rate": first_attempt_successes / completed if completed > 0 else 0.0,
        "avg_attempts": total_attempts / completed if completed > 0 else 0.0,
        "tokens_per_task": total_tokens / completed if completed > 0 else 0,
        "cost_per_task": total_cost / completed if completed > 0 else 0.0,
        "quality_pass_rate": quality_full_pass / completed if completed > 0 else 0.0,
        "functional_pass_rate": criteria_pass / total_criteria if total_criteria > 0 else 0.0,
        "test_edge_case_rate": edge_cases_pass / tasks_with_tests if tasks_with_tests > 0 else 0.0,
        "completed_count": completed,
        "failed_count": failed,
        "total_tokens": total_tokens,
        "total_cost_usd": total_cost,
        "criteria_pass": criteria_pass,
        "criteria_partial": criteria_partial,
        "criteria_fail": criteria_fail,
        "total_criteria": total_criteria,
        "quality_dimensions": quality_dimensions,
        "quality_totals": quality_totals,
    }


def get_failed_tasks(state: dict) -> list[dict]:
    """Get details of failed tasks."""
    failed = []
    for tid, task in state.get("tasks", {}).items():
        if task.get("status") == "failed":
            failed.append({
                "id": tid,
                "name": task.get("name", ""),
                "error": task.get("error", "Unknown error"),
            })
    return failed


def get_improvement_patterns(state: dict) -> list[str]:
    """Identify improvement patterns from verification data."""
    patterns = []
    edge_case_issues = 0
    type_issues = 0
    doc_issues = 0

    for task in state.get("tasks", {}).values():
        verification = task.get("verification", {})

        tests = verification.get("tests", {})
        if tests.get("edge_cases") in ["PARTIAL", "FAIL"]:
            edge_case_issues += 1

        quality = verification.get("quality", {})
        if quality.get("types") in ["PARTIAL", "FAIL"]:
            type_issues += 1
        if quality.get("docs") in ["PARTIAL", "FAIL"]:
            doc_issues += 1

    if edge_case_issues > 0:
        patterns.append(f"{edge_case_issues} task(s) had issues with edge case testing")
    if type_issues > 0:
        patterns.append(f"{type_issues} task(s) had type annotation issues")
    if doc_issues > 0:
        patterns.append(f"{doc_issues} task(s) had documentation issues")

    return patterns


def print_report(state: dict, include_validation: bool = False) -> None:
    """Print full evaluation report."""
    metrics = compute_metrics(state)
    tasks = state.get("tasks", {})
    artifacts = state.get("artifacts", {})

    # Header
    print("Execution Evaluation Report")
    print("=" * 60)
    print()

    # Planning Quality
    task_validation = artifacts.get("task_validation", {})
    print("Planning Quality")
    print("-" * 40)
    verdict = task_validation.get("verdict", "N/A")
    issues = task_validation.get("issues", [])
    print(f"Plan Verdict: {verdict}")
    print(f"Issues at Planning: {len(issues)}")
    print()

    # Execution Summary
    print("Execution Summary")
    print("-" * 40)
    total_tasks = len(tasks)
    completed = metrics["completed_count"]
    failed = metrics["failed_count"]
    blocked = sum(1 for t in tasks.values() if t.get("status") == "blocked")
    skipped = sum(1 for t in tasks.values() if t.get("status") == "skipped")

    print(f"Tasks: {total_tasks} total")
    if completed > 0:
        print(f"  Completed:     {completed} ({completed/total_tasks*100:.0f}%)" if total_tasks > 0 else f"  Completed:     {completed}")
    if failed > 0:
        print(f"  Failed:        {failed} ({failed/total_tasks*100:.0f}%)" if total_tasks > 0 else f"  Failed:        {failed}")
    if blocked > 0:
        print(f"  Blocked:       {blocked} ({blocked/total_tasks*100:.0f}%)" if total_tasks > 0 else f"  Blocked:       {blocked}")
    if skipped > 0:
        print(f"  Skipped:       {skipped} ({skipped/total_tasks*100:.0f}%)" if total_tasks > 0 else f"  Skipped:       {skipped}")
    print()

    # First-attempt metrics
    first_attempt = int(metrics["first_attempt_success_rate"] * completed) if completed > 0 else 0
    print(f"First-Attempt Success: {first_attempt}/{completed} ({metrics['first_attempt_success_rate']:.0%})")
    print(f"Average Attempts: {metrics['avg_attempts']:.2f}")
    print()

    # Verification Breakdown
    print("Verification Breakdown")
    print("-" * 40)

    # Functional criteria
    print("Functional Criteria:")
    total_c = metrics["total_criteria"]
    if total_c > 0:
        print(f"  PASS:     {metrics['criteria_pass']}/{total_c} ({metrics['criteria_pass']/total_c*100:.0f}%)")
        print(f"  PARTIAL:  {metrics['criteria_partial']}/{total_c} ({metrics['criteria_partial']/total_c*100:.0f}%)")
        print(f"  FAIL:     {metrics['criteria_fail']}/{total_c} ({metrics['criteria_fail']/total_c*100:.0f}%)")
    else:
        print("  No criteria data")
    print()

    # Code quality
    print("Code Quality:")
    qd = metrics["quality_dimensions"]
    qt = metrics["quality_totals"]
    for dim in ["types", "docs", "patterns", "errors"]:
        total = qt[dim]
        passed = qd[dim]
        if total > 0:
            print(f"  {dim.capitalize():10} {passed}/{total} PASS")
    print()

    # Cost Analysis
    print("Cost Analysis")
    print("-" * 40)
    print(f"Total Tokens:  {metrics['total_tokens']:,}")
    print(f"Total Cost:    ${metrics['total_cost_usd']:.2f}")
    if completed > 0:
        print(f"Per Task:      ${metrics['cost_per_task']:.4f}")
    print()

    # Failure Analysis
    failed_tasks = get_failed_tasks(state)
    if failed_tasks:
        print("Failure Analysis")
        print("-" * 40)
        for ft in failed_tasks:
            print(f"{ft['id']}: FAIL - {ft['error']}")
        print()

    # Improvement Patterns
    patterns = get_improvement_patterns(state)
    if patterns:
        print("Improvement Patterns")
        print("-" * 40)
        for p in patterns:
            print(f"- {p}")
        print()

    # Validation Section (if requested and available)
    if include_validation and VALIDATION_AVAILABLE:
        print("Workflow Validation")
        print("-" * 40)

        # DAG validation
        dag_valid, dag_msg = validate_dag(state)
        if dag_valid:
            print("✓ DAG: No dependency cycles")
        else:
            print(f"✗ DAG: {dag_msg}")

        # Steel thread validation
        st_valid, st_issues = validate_steel_thread(state)
        if st_valid:
            print("✓ Steel Thread: Valid path")
        else:
            print("✗ Steel Thread issues:")
            for issue in st_issues:
                print(f"    - {issue}")

        # Verification commands
        vc_valid, vc_issues = validate_all_verification_commands(state)
        if vc_valid:
            print("✓ Verification Commands: All valid")
        else:
            print(f"✗ Verification Commands: {len(vc_issues)} task(s) with issues")

        # Verifier calibration
        cal = compute_calibration_metrics(state)
        print(f"\nVerifier Calibration Score: {cal['calibration_score']:.1%}")
        if cal["false_positives"]:
            print(f"  False Positives: {len(cal['false_positives'])}")
        if cal["false_negatives"]:
            print(f"  Potential False Negatives: {len(cal['false_negatives'])}")
        print()
    elif include_validation and not VALIDATION_AVAILABLE:
        print("Validation module not available")
        print()


def print_json(state: dict, include_validation: bool = False) -> None:
    """Print evaluation data as JSON."""
    metrics = compute_metrics(state)
    failed_tasks = get_failed_tasks(state)
    patterns = get_improvement_patterns(state)
    artifacts = state.get("artifacts", {})
    task_validation = artifacts.get("task_validation", {})

    output = {
        "planning": {
            "verdict": task_validation.get("verdict"),
            "issues_count": len(task_validation.get("issues", [])),
        },
        "metrics": metrics,
        "failed_tasks": failed_tasks,
        "improvement_patterns": patterns,
    }

    # Add validation if requested
    if include_validation and VALIDATION_AVAILABLE:
        validation = run_all_validations(state)
        output["validation"] = {
            "dag_valid": validation["dag"]["valid"],
            "steel_thread_valid": validation["steel_thread"]["valid"],
            "verification_commands_valid": validation["verification_commands"]["valid"],
            "calibration": validation["calibration"],
            "overall_valid": validation["overall_valid"],
        }

    print(json.dumps(output, indent=2))


def print_metrics_only(state: dict) -> None:
    """Print only computed metrics."""
    metrics = compute_metrics(state)
    # Remove internal counters from output
    output_metrics = {
        "task_success_rate": metrics["task_success_rate"],
        "first_attempt_success_rate": metrics["first_attempt_success_rate"],
        "avg_attempts": metrics["avg_attempts"],
        "tokens_per_task": metrics["tokens_per_task"],
        "cost_per_task": metrics["cost_per_task"],
        "quality_pass_rate": metrics["quality_pass_rate"],
        "functional_pass_rate": metrics["functional_pass_rate"],
        "test_edge_case_rate": metrics["test_edge_case_rate"],
    }
    print(json.dumps(output_metrics, indent=2))


def main() -> None:
    state = load_state()
    if not state:
        print("No state file found. Run 'state.py init <target_dir>' first.", file=sys.stderr)
        sys.exit(1)

    output_format = "text"
    metrics_only = False
    include_validation = False

    args = sys.argv[1:]
    i = 0
    while i < len(args):
        if args[i] == "--format" and i + 1 < len(args):
            output_format = args[i + 1]
            i += 2
        elif args[i] == "--metrics-only":
            metrics_only = True
            i += 1
        elif args[i] == "--validate":
            include_validation = True
            i += 1
        else:
            i += 1

    if metrics_only:
        print_metrics_only(state)
    elif output_format == "json":
        print_json(state, include_validation)
    else:
        print_report(state, include_validation)


if __name__ == "__main__":
    main()
